File: 90_Subsets_II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
	if(nums.length===0)return [[]];
  nums.sort();
  let results = [[],[nums[0]]];
  let s = new WeakSet();
  s.add([nums[0]])
  let p = nums[1] === nums[0]?1:0;

  for(let i =1;i<nums.length;i++){
  	if(nums[i]===nums[i-1]){
  		const l = results.length;
  		for(let k=p;k<l;k++){
  			results.push(results[k].concat(nums[i]));
  		}
  		p = l;
  	}
  	else{
  		p = results.length;
	  	for(let j in results){
	  		let temp = results[j].concat(nums[i]);
	  		results.push(temp);
	  	}
	  }
  }
  return results;
};

File: 90.Subsets_II.py
class Solution:
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        answer = []
        nums.sort()
        def backtrack(arr, index):
            if index < len(nums):
                answer.append(arr)
                if index < len(nums):
                    for i, ele in enumerate(nums[index+1:]):
                        if i != 0 and ele == nums[index+i]:
                            continue
                        else:
                            backtrack(arr[:]+[ele], index+i+1)
                else:
                    return
        backtrack([], -1)
        return answer

print(Solution().subsetsWithDup([1, 2, 2,2]))
